### PR TITLE
Custom model support

### PR DESCRIPTION
**Pull request by frankperez87**
Work in progress for supporting custom models. 

The idea is to allow a user to define their own models to be used with this package based on what is set in the config. 

**What works now with this PR**
- You can now specify custom models to be used via config.
- Updated Readme to show new configuration options.

**Whats Left**
- Writing automated tests.

### CODE REVIEW

**Comment by karunkshrestha** [comment]
I have been waiting for this feature to allow me to set a different db instance/tablename. It would be great addition once the merge request is approved.

File: src/Contracts/MonitoredScheduledTask.php
Comment (freekmurze) [inline src/Contracts/MonitoredScheduledTask.php:13]: Instead of interface, I would like it better if, everywhere we need the models, we would check if the model given is or extends the one that ships with the package. 

This way people could still override all functions by providing a custom model.  And as a package maintainer we would be able to add extra functions in a non-breaking way.

**Review by freekmurze — CHANGES_REQUESTED** [review]
Thanks for your work on this.

File: src/Contracts/MonitoredScheduledTask.php
Comment (frankperez87) [reply src/Contracts/MonitoredScheduledTask.php:13]: @freekmurze No problem. The thought here was when I was using it locally, we could just extend the existing models to override anything needed but at least they would still be constrained to the contract :) 

For example, in my use case I simply did this to update the database connection.
```php
class MonitoredScheduledTask extends ModelsMonitoredScheduledTask
{
    protected $connection = 'log';
}
```

Let me know what you think about that, or if you'd still prefer changing everything to add those conditional checks. I typically prefer Interfaces and binding it from a service provider, but I'm definitely open to trying a different approach.

Comment (freekmurze) [reply src/Contracts/MonitoredScheduledTask.php:13]: > Let me know what you think about that, or if you'd still prefer changing everything to add those conditional checks. I typically prefer Interfaces and binding it from a service provider, but I'm definitely open to trying a different approach.

In a project context I like this too, but in a package context I tend to avoid interfaces unless it's a really good fit. 

Let's do the conditional check!

Comment (frankperez87) [reply src/Contracts/MonitoredScheduledTask.php]: Hey @freekmurze I went ahead and implemented a new solution where we now have a method within the service provider that checks to make sure the custom model extends our package model :) 

I think this might be a much cleaner way of implementing the check compared to adding conditionals everywhere we need the models. 

What do you think?

File: src/ScheduleMonitorServiceProvider.php
Comment (frankperez87) [inline src/ScheduleMonitorServiceProvider.php:62]:
```suggestion
            if (! class_exists('CreateScheduleMonitorTables')) {
```

File: src/ScheduleMonitorServiceProvider.php
Comment (frankperez87) [inline src/ScheduleMonitorServiceProvider.php:86]:
```suggestion
        if (! class_exists(OhDear::class)) {
```

File: src/Commands/SyncCommand.php
Comment (freekmurze) [inline src/Commands/SyncCommand.php:41]: To me, using `::` twice in a statement hurts readability.

Maybe you could add a trait `UsesScheduleMonitoringModels` that has a function `getMonitoredScheduleTaskModel` and use that wherever you need to use that model.

**Review by freekmurze — CHANGES_REQUESTED** [review]

File: src/Contracts/MonitoredScheduledTask.php
Comment (freekmurze) [reply src/Contracts/MonitoredScheduledTask.php]: Thanks for your work. The implementation can be improved imho, but you could already write some tests around it so things can be refactored safely.

File: src/Commands/SyncCommand.php
Comment (frankperez87) [reply src/Commands/SyncCommand.php:41]: @freekmurze Sounds good. Great suggestion. 👍 I'll get that implemented.

File: src/Contracts/MonitoredScheduledTask.php
Comment (frankperez87) [reply src/Contracts/MonitoredScheduledTask.php]: @freekmurze I'll work on getting the tests written 👍

File: src/Commands/SyncCommand.php
Comment (frankperez87) [reply src/Commands/SyncCommand.php]: @freekmurze I've gone ahead and changed out the references to the classes to now use a trait.  I've also added a test that checks that the trait methods return instances of the expected models.

**Comment by frankperez87** [comment]
@freekmurze One thing I noticed, when running `composer format` on the codebase it seemed like it corrected quite a bit. I just want to confirm that the formatting set now is what is expected on your end?

**Comment by frankperez87** [comment]
Hey @freekmurze I also have a question for you regarding testing expectations that we're throwing in the service provider. 

Currently, if a class defined in the config is not extending one of our package classes we throw an exception to let the user know. Testing this seems to be a little tricky though with workbench. Do you have any experience testing a case like this? 

What I was trying to do was something like the following but I can't seem to catch the exception.
```php
    protected function setInvalidModelBindings($app)
    {
        $invalidClass = new class {};

        $app->config->set(
            'schedule-monitor.models.monitored_scheduled_task',
            get_class($invalidClass)
        );
    }

    /** 
     * @test 
     * @define-env setInvalidModelBindings
     * */
    public function it_can_protect_against_invalid_model_bindings()
    {
        $this->expectException(InvalidClassException::class);
    }
```

**Comment by freekmurze** [comment]
Implementation looks good, I'll take care of tests later 👍 

Thanks for your work on this!

**Comment by karunkshrestha** [comment]
This is great. Thank you @frankperez87 @freekmurze

**Pull request merged**
